sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala

```diff
@@ -769,4 +769,33 @@ class ParametersSuite extends QueryTest with SharedSparkSession with PlanTest {
     checkAnswer(spark.sql(query(":cte"), args = Map("cte" -> "t1")), Row(1))
     checkAnswer(spark.sql(query("?"), args = Array("t1")), Row(1))
   }
+
+  test("SPARK-50403: parameterized execute immediate") {
+    checkAnswer(spark.sql("execute immediate 'select ?' using ?", Array(1)), Row(1))
+    checkAnswer(spark.sql("execute immediate 'select ?, ?' using ?, 2", Array(1)), Row(1, 2))
+    checkError(
+      exception = intercept[AnalysisException] {
+        spark.sql("execute immediate 'select ?, ?' using 1", Array(2))
+      },
+      condition = "UNBOUND_SQL_PARAMETER",
+      parameters = Map("name" -> "_10"),
+      context = ExpectedContext("?", 10, 10))
+
+    checkAnswer(spark.sql("execute immediate 'select :param1' using :param2 as param1",
+      Map("param2" -> 42)), Row(42))
+    checkAnswer(spark.sql(
+      "execute immediate 'select :param1, :param2' using :param2 as param1, 43 as param2",
+      Map("param2" -> 42)), Row(42, 43))
+    checkError(
+      exception = intercept[AnalysisException] {
+        spark.sql("execute immediate 'select :param1, :param2' using 1 as param1",
+          Map("param2" -> 2))
+      },
+      condition = "UNBOUND_SQL_PARAMETER",
+      parameters = Map("name" -> "param2"),
+      context = ExpectedContext(":param2", 16, 22))
+
+    checkAnswer(spark.sql("execute immediate 'select ?' using :param", Map("param" -> 2)), Row(2))
+    checkAnswer(spark.sql("execute immediate 'select :param' using ? as param", Array(3)), Row(3))
```

Comment:
LGTM,

maybe add one error test case where you bind in a named parameter that matches in the string, but not the USING.
That should be ignored (IIRC that we can bind in parameter markers that are not consumed)

Reply:
Isn't the case (it presents already). `param2` matches in the string but not the USING, and we get the error `UNBOUND_SQL_PARAMETER`:
```scala
spark.sql(
      "execute immediate 'select :param1, :param2' using :param2 as param1, 43 as param2",
      Map("param2" -> 42))
```
If not, could you give an example.

Reply:
Just in case, going to add this test case:
```scala
    checkAnswer(spark.sql("execute immediate 'select :param' using 0 as param",
      Map("param" -> 42)), Row(0))
```